### PR TITLE
Add on chain for federations

### PR DESCRIPTION
**Pull request by TonyGiorgio**
TODO:
- [x] Withdrawals
- [x] Frontend needs to allow withdrawals from fedimint balance
- [x] Change APIs to account for withdraw pending
- [x] Figure out max send situation
- [x] Monitor eventual TXID for confirmations (fedimint doesn't care about tracking it)
- [x] Lots of TODOs in the on chain cleanup in federation.rs
- [x] pull activity list for on chain addresses
- [x] Refreshing doesn't show on chain history
- [x] activity item for on chain deposit doesn't show additional info
- [x] tests for activity item & indexing
- [x] subscribe to pending on chain operations
- [x] show payment details on the onchain send screen when the link is clicked
- [x] handle tiny amounts properly if they are attempted to be paid

### CODE REVIEW

**Comment by TonyGiorgio** [comment]
This works as long as you don't try to send max. Still a lot of code to clean up and want to do some tests.

There's some possible ideas for frontend changes too but will think about them more. 

- we should say not to send more than once to the address
- we should tell them it requires 10 confs for deposits
- we should probably show a fee to confirm before sending

**Comment by TonyGiorgio** [comment]
I want to do some tests for the activity item stuff but this is ready for review

**Review by benthecarman — COMMENTED** [review]
overall looks good

File: mutiny-core/src/federation.rs [hunk 1]
```diff
@@ -630,6 +684,105 @@ impl<S: MutinyStorage> FederationClient<S> {
...
+        // keep subscribing if txid wasn't retrieved, but then return timeout
+        let operation = self
+            .fedimint_client
+            .operation_log()
+            .get_operation(op_id)
+            .await
+            .expect("just created it");
+        self.subscribe_operation(operation, op_id);
+
+        Err(MutinyError::PaymentTimeout)
```
[the remark this answers is not in the document]
Comment (TonyGiorgio) [reply mutiny-core/src/federation.rs:764]: I feel that, but it should have the transaction showing up in the pending list for them to see is in progress.

File: mutiny-core/src/storage.rs [hunk 2]
```diff
@@ -881,6 +883,80 @@ impl MutinyStorage for () {
...
+                timestamp: None,
+                key,
+            });
+        }
+    }
+
+    Ok(())
+}
+
+pub(crate) fn delete_transaction_details<S: MutinyStorage>(
```
[the remark this answers is not in the document]
Comment (TonyGiorgio) [reply mutiny-core/src/storage.rs:929]: it deletes it no matter, it's just the index whether or not it is removed

**Comment by TonyGiorgio** [comment]
READY

File: mutiny-core/src/federation.rs [hunk 3]
```diff
@@ -1087,6 +1321,277 @@ where
...
+                break;
+            };
+
+            match esplora.get_tx_status(&txid).await {
+                Ok(s) => {
+                    if s.confirmed {
+                        log_info!(logger, "Transaction confirmed");
+                        transaction_details.confirmation_time = ConfirmationTime::Confirmed {
+                            height: s.block_height.expect("confirmed"),
+                            time: now().as_secs(),
```
Comment (benthecarman) [inline mutiny-core/src/federation.rs:1448]: should use the block's timestamp here if that's available, that's what BDK does and will keep things consistent

File: mutiny-core/src/federation.rs [hunk 4]
```diff
@@ -1087,6 +1321,277 @@ where
...
+                            }
+                        }
+                    }
+                }
+                Err(e) => {
+                    log_error!(logger, "Error updating transaction: {e}");
+                }
+            }
+
+            sleep(5_000).await;
```
Comment (benthecarman) [inline mutiny-core/src/federation.rs:1466]: should do this every minute or something, 5 seconds is really low and will spam mempool.space

**Review by benthecarman — COMMENTED** [review]
two small things, otherwise LGTM, nice job

**Review by benthecarman — APPROVED** [review]
:ship: it

**Pull request merged**